From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sameer Pujar <spujar@nvidia.com>
Date: Wed, 13 Mar 2024 18:08:57 +0000
Subject: [PATCH 211/243] NVIDIA: SAUCE: ASoC: tegra: codec2codec mode for AHUB
 DAI links

Add support to use AHUB internal modules with codec2codec DAI links.
This can be enabled by using DT property "nvidia,ahub-c2c-links" in
the sound node. The FE DAI links use normal PCM interfaces and thus
are normal DAI links. Rest of the AHUB modules are connected via
codec2codec DAI links. This is enabled with graph card driver now.

Bug 4451662

Change-Id: I8d2b78805dfa03b7ec5d53747c583d60d5ab36de
Signed-off-by: Sameer Pujar <spujar@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/3102888
GVS: Gerrit_Virtual_Submit <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Sharad Gupta <sharadg@nvidia.com>
Reviewed-by: Mohan kumar <mkumard@nvidia.com>

diff --git a/include/sound/simple_card_utils.h b/include/sound/simple_card_utils.h
index 69a9c9c4d0e9c..5646c14f6c79e 100644
--- a/include/sound/simple_card_utils.h
+++ b/include/sound/simple_card_utils.h
@@ -70,6 +70,7 @@ struct simple_util_priv {
 	struct snd_soc_dai_link *dai_link;
 	struct simple_util_dai *dais;
 	struct snd_soc_dai_link_component *dlcs;
+	struct snd_soc_dai_link_component *dummy_dlc;
 	struct snd_soc_codec_conf *codec_conf;
 	struct gpio_desc *pa_gpio;
 	const struct snd_soc_ops *ops;
diff --git a/sound/soc/generic/audio-graph-card.c b/sound/soc/generic/audio-graph-card.c
index a8a3bad3df00d..15d4902734766 100644
--- a/sound/soc/generic/audio-graph-card.c
+++ b/sound/soc/generic/audio-graph-card.c
@@ -297,12 +297,18 @@ static int graph_dai_link_of(struct simple_util_priv *priv,
 	if (ret < 0)
 		goto end;
 
+	if (!codec_ep) {
+		codecs = priv->dummy_dlc;
+		dai_link->codecs = priv->dummy_dlc;
+	}
+
 	ret = graph_parse_node(priv, codec_ep, li, NULL);
 	if (ret < 0)
 		goto end;
 
 	snprintf(dai_name, sizeof(dai_name),
-		 "%s-%s", cpus->dai_name, codecs->dai_name);
+		 "%pOFP.%s-%pOFP.%s", dai_link->cpus->of_node, cpus->dai_name,
+		 dai_link->codecs->of_node, codecs->dai_name);
 
 	simple_util_canonicalize_cpu(cpus, is_single_links);
 	simple_util_canonicalize_platform(platforms, cpus);
diff --git a/sound/soc/generic/simple-card-utils.c b/sound/soc/generic/simple-card-utils.c
index 9e5be0eaa77f3..e6ca99941128a 100644
--- a/sound/soc/generic/simple-card-utils.c
+++ b/sound/soc/generic/simple-card-utils.c
@@ -265,6 +265,9 @@ int simple_util_parse_clk(struct device *dev,
 	struct clk *clk;
 	u32 val;
 
+	if (!node)
+		return 0;
+
 	/*
 	 * Parse dai->sysclk come from "clocks = <&xxx>"
 	 * (if system has common clock)
diff --git a/sound/soc/soc-link.c b/sound/soc/soc-link.c
index 02fd68f2e7028..8415173c7a0e8 100644
--- a/sound/soc/soc-link.c
+++ b/sound/soc/soc-link.c
@@ -50,6 +50,7 @@ int snd_soc_link_be_hw_params_fixup(struct snd_soc_pcm_runtime *rtd,
 
 	return soc_link_ret(rtd, ret);
 }
+EXPORT_SYMBOL_GPL(snd_soc_link_be_hw_params_fixup);
 
 int snd_soc_link_startup(struct snd_pcm_substream *substream)
 {
diff --git a/sound/soc/tegra/tegra186_asrc.c b/sound/soc/tegra/tegra186_asrc.c
index c0851fe5c45b1..6a482c04c59bc 100644
--- a/sound/soc/tegra/tegra186_asrc.c
+++ b/sound/soc/tegra/tegra186_asrc.c
@@ -927,6 +927,7 @@ static const struct snd_soc_component_driver tegra186_asrc_cmpnt = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra186_asrc_widgets),
 	.controls		= tegra186_asrc_controls,
 	.num_controls		= ARRAY_SIZE(tegra186_asrc_controls),
+	.endianness		= 1,
 };
 
 static bool tegra186_asrc_wr_reg(struct device *dev, unsigned int reg)
diff --git a/sound/soc/tegra/tegra186_dspk.c b/sound/soc/tegra/tegra186_dspk.c
index 08b629cb5c3a6..891bf5af5931b 100644
--- a/sound/soc/tegra/tegra186_dspk.c
+++ b/sound/soc/tegra/tegra186_dspk.c
@@ -450,6 +450,7 @@ static const struct snd_soc_component_driver tegra186_dspk_cmpnt = {
 	.num_dapm_widgets = ARRAY_SIZE(tegra186_dspk_widgets),
 	.controls = tegrat186_dspk_controls,
 	.num_controls = ARRAY_SIZE(tegrat186_dspk_controls),
+	.endianness = 1,
 };
 
 static bool tegra186_dspk_wr_reg(struct device *dev, unsigned int reg)
diff --git a/sound/soc/tegra/tegra210_adx.c b/sound/soc/tegra/tegra210_adx.c
index bbdd079875229..5f917916b57a2 100644
--- a/sound/soc/tegra/tegra210_adx.c
+++ b/sound/soc/tegra/tegra210_adx.c
@@ -547,6 +547,7 @@ static const struct snd_soc_component_driver tegra210_adx_cmpnt = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra210_adx_widgets),
 	.controls		= tegra210_adx_controls,
 	.num_controls		= ARRAY_SIZE(tegra210_adx_controls),
+	.endianness		= 1,
 };
 
 static bool tegra210_adx_wr_reg(struct device *dev,
diff --git a/sound/soc/tegra/tegra210_ahub.c b/sound/soc/tegra/tegra210_ahub.c
index ecea0cbfbb1bc..2d3ad8df60e85 100644
--- a/sound/soc/tegra/tegra210_ahub.c
+++ b/sound/soc/tegra/tegra210_ahub.c
@@ -2026,6 +2026,7 @@ static const struct snd_soc_component_driver tegra210_ahub_component = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra210_ahub_widgets),
 	.dapm_routes		= tegra210_ahub_routes,
 	.num_dapm_routes	= ARRAY_SIZE(tegra210_ahub_routes),
+	.endianness		= 1,
 };
 
 static const struct snd_soc_component_driver tegra186_ahub_component = {
@@ -2033,6 +2034,7 @@ static const struct snd_soc_component_driver tegra186_ahub_component = {
 	.num_dapm_widgets = ARRAY_SIZE(tegra186_ahub_widgets),
 	.dapm_routes = tegra186_ahub_routes,
 	.num_dapm_routes = ARRAY_SIZE(tegra186_ahub_routes),
+	.endianness = 1,
 };
 
 static const struct snd_soc_component_driver tegra234_ahub_component = {
@@ -2040,6 +2042,7 @@ static const struct snd_soc_component_driver tegra234_ahub_component = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra234_ahub_widgets),
 	.dapm_routes		= tegra186_ahub_routes,
 	.num_dapm_routes	= ARRAY_SIZE(tegra186_ahub_routes),
+	.endianness		= 1,
 };
 
 static const struct snd_soc_component_driver tegra264_ahub_component = {
@@ -2047,6 +2050,7 @@ static const struct snd_soc_component_driver tegra264_ahub_component = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra264_ahub_widgets),
 	.dapm_routes		= tegra264_ahub_routes,
 	.num_dapm_routes	= ARRAY_SIZE(tegra264_ahub_routes),
+	.endianness		= 1,
 };
 
 static bool tegra210_ahub_wr_reg(struct device *dev, unsigned int reg)
diff --git a/sound/soc/tegra/tegra210_amx.c b/sound/soc/tegra/tegra210_amx.c
index 82d64f868ccf8..c97f7ee2e3606 100644
--- a/sound/soc/tegra/tegra210_amx.c
+++ b/sound/soc/tegra/tegra210_amx.c
@@ -556,6 +556,7 @@ static const struct snd_soc_component_driver tegra210_amx_cmpnt = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra210_amx_widgets),
 	.controls		= tegra210_amx_controls,
 	.num_controls		= ARRAY_SIZE(tegra210_amx_controls),
+	.endianness		= 1,
 };
 
 static bool tegra210_amx_wr_reg(struct device *dev, unsigned int reg)
diff --git a/sound/soc/tegra/tegra210_dmic.c b/sound/soc/tegra/tegra210_dmic.c
index 6fd4f05124ea9..9be1e1ac9746e 100644
--- a/sound/soc/tegra/tegra210_dmic.c
+++ b/sound/soc/tegra/tegra210_dmic.c
@@ -461,6 +461,7 @@ static const struct snd_soc_component_driver tegra210_dmic_compnt = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra210_dmic_widgets),
 	.controls		= tegra210_dmic_controls,
 	.num_controls		= ARRAY_SIZE(tegra210_dmic_controls),
+	.endianness		= 1,
 };
 
 static bool tegra210_dmic_wr_reg(struct device *dev, unsigned int reg)
diff --git a/sound/soc/tegra/tegra210_i2s.c b/sound/soc/tegra/tegra210_i2s.c
index 980e820b523bf..7c5a6e6c89305 100644
--- a/sound/soc/tegra/tegra210_i2s.c
+++ b/sound/soc/tegra/tegra210_i2s.c
@@ -899,15 +899,16 @@ static const struct snd_soc_component_driver tegra210_i2s_cmpnt = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra210_i2s_widgets),
 	.controls		= tegra210_i2s_controls,
 	.num_controls		= ARRAY_SIZE(tegra210_i2s_controls),
+	.endianness		= 1,
 };
 
 static const struct snd_soc_component_driver tegra264_i2s_cmpnt = {
+	.probe			= tegra210_i2s_component_probe,
 	.dapm_widgets		= tegra264_i2s_widgets,
 	.num_dapm_widgets	= ARRAY_SIZE(tegra264_i2s_widgets),
-	.dapm_routes		= tegra210_i2s_routes,
-	.num_dapm_routes	= ARRAY_SIZE(tegra210_i2s_routes),
 	.controls		= tegra210_i2s_controls,
 	.num_controls		= ARRAY_SIZE(tegra210_i2s_controls),
+	.endianness		= 1,
 };
 
 static bool tegra210_i2s_wr_reg(struct device *dev, unsigned int reg)
diff --git a/sound/soc/tegra/tegra210_mixer.c b/sound/soc/tegra/tegra210_mixer.c
index 12753614aa0fa..83ca0f783e1f3 100644
--- a/sound/soc/tegra/tegra210_mixer.c
+++ b/sound/soc/tegra/tegra210_mixer.c
@@ -833,6 +833,7 @@ static const struct snd_soc_component_driver tegra210_mixer_cmpnt = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra210_mixer_widgets),
 	.controls		= tegra210_mixer_gain_ctls,
 	.num_controls		= ARRAY_SIZE(tegra210_mixer_gain_ctls),
+	.endianness		= 1,
 };
 
 static bool tegra210_mixer_wr_reg(struct device *dev,
diff --git a/sound/soc/tegra/tegra210_mvc.c b/sound/soc/tegra/tegra210_mvc.c
index c76fdf6e85b56..5b2aad0220aa4 100644
--- a/sound/soc/tegra/tegra210_mvc.c
+++ b/sound/soc/tegra/tegra210_mvc.c
@@ -667,6 +667,7 @@ static const struct snd_soc_component_driver tegra210_mvc_cmpnt = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra210_mvc_widgets),
 	.controls		= tegra210_mvc_vol_ctrl,
 	.num_controls		= ARRAY_SIZE(tegra210_mvc_vol_ctrl),
+	.endianness		= 1,
 };
 
 static bool tegra210_mvc_rd_reg(struct device *dev, unsigned int reg)
diff --git a/sound/soc/tegra/tegra210_ope.c b/sound/soc/tegra/tegra210_ope.c
index 59b446b7c773e..a62ad268b9550 100644
--- a/sound/soc/tegra/tegra210_ope.c
+++ b/sound/soc/tegra/tegra210_ope.c
@@ -259,6 +259,7 @@ static const struct snd_soc_component_driver tegra210_ope_cmpnt = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra210_ope_widgets),
 	.controls		= tegra210_ope_controls,
 	.num_controls		= ARRAY_SIZE(tegra210_ope_controls),
+	.endianness		= 1,
 };
 
 static bool tegra210_ope_wr_reg(struct device *dev, unsigned int reg)
diff --git a/sound/soc/tegra/tegra210_sfc.c b/sound/soc/tegra/tegra210_sfc.c
index 925e2f3a7637a..c4d42a577074b 100644
--- a/sound/soc/tegra/tegra210_sfc.c
+++ b/sound/soc/tegra/tegra210_sfc.c
@@ -3526,6 +3526,7 @@ static const struct snd_soc_component_driver tegra210_sfc_cmpnt = {
 	.num_dapm_widgets	= ARRAY_SIZE(tegra210_sfc_widgets),
 	.controls		= tegra210_sfc_controls,
 	.num_controls		= ARRAY_SIZE(tegra210_sfc_controls),
+	.endianness		= 1,
 };
 
 static bool tegra210_sfc_wr_reg(struct device *dev, unsigned int reg)
diff --git a/sound/soc/tegra/tegra_audio_graph_card.c b/sound/soc/tegra/tegra_audio_graph_card.c
index b93a61db9ed0b..beca869559887 100644
--- a/sound/soc/tegra/tegra_audio_graph_card.c
+++ b/sound/soc/tegra/tegra_audio_graph_card.c
@@ -6,10 +6,12 @@
 #include <linux/math64.h>
 #include <linux/module.h>
 #include <linux/of.h>
+#include <linux/of_platform.h>
 #include <linux/platform_device.h>
 #include <sound/graph_card.h>
 #include <sound/pcm_params.h>
 #include <sound/soc-dai.h>
+#include <sound/soc-link.h>
 
 #define MAX_PLLA_OUT0_DIV 128
 
@@ -153,6 +155,9 @@ static int tegra_audio_graph_hw_params(struct snd_pcm_substream *substream,
 {
 	struct snd_soc_pcm_runtime *rtd = snd_soc_substream_to_rtd(substream);
 	struct snd_soc_dai *cpu_dai = snd_soc_rtd_to_cpu(rtd, 0);
+	struct snd_soc_pcm_stream *dai_params;
+	struct snd_soc_pcm_runtime *vrtd;
+	struct snd_pcm_hw_params codec_params;
 	int err;
 
 	if (need_clk_update(cpu_dai)) {
@@ -161,6 +166,34 @@ static int tegra_audio_graph_hw_params(struct snd_pcm_substream *substream,
 			return err;
 	}
 
+	/*
+	 * For AHUB in c2c mode, hw_params() call here happens only for
+	 * FE links. The PLL update is required to take care of I/O
+	 * configuration.
+	 *
+	 */
+	if (cpu_dai->driver->ops && !rtd->card->component_chaining) {
+		for_each_card_rtds(rtd->card, vrtd) {
+			if (!vrtd->dai_link->c2c_params)
+				continue;
+
+			codec_params = *params;
+			snd_soc_link_be_hw_params_fixup(vrtd, &codec_params);
+
+			dai_params = (struct snd_soc_pcm_stream *)vrtd->dai_link->c2c_params;
+			dai_params->rate_min = params_rate(&codec_params);
+			dai_params->channels_min = params_channels(&codec_params);
+			dai_params->formats = 1 << params_format(&codec_params);
+
+			cpu_dai = snd_soc_rtd_to_cpu(vrtd, 0);
+			if (need_clk_update(cpu_dai)) {
+				err = tegra_audio_graph_update_pll(substream, &codec_params);
+				if (err)
+					return err;
+			}
+		}
+	}
+
 	return simple_util_hw_params(substream, params);
 }
 
@@ -193,11 +226,58 @@ static int tegra_audio_graph_card_probe(struct snd_soc_card *card)
 	return ret;
 }
 
+static struct snd_soc_dai_driver tegra_dummy_dai = {
+	.name = "tegra-snd-dummy-dai",
+	.playback = {
+		.stream_name	= "Dummy Playback",
+		.channels_min	= 1,
+		.channels_max	= 32,
+		.rates		= SNDRV_PCM_RATE_8000_192000,
+		.formats	= SNDRV_PCM_FMTBIT_S8 |
+				  SNDRV_PCM_FMTBIT_S16_LE |
+				  SNDRV_PCM_FMTBIT_S32_LE,
+	},
+	.capture = {
+		.stream_name	= "Dummy Capture",
+		.channels_min	= 1,
+		.channels_max	= 32,
+		.rates		= SNDRV_PCM_RATE_8000_192000,
+		.formats	= SNDRV_PCM_FMTBIT_S8 |
+				  SNDRV_PCM_FMTBIT_S16_LE |
+				  SNDRV_PCM_FMTBIT_S32_LE,
+	 },
+};
+
+static const struct snd_soc_dapm_widget tegra_dummy_widgets[] = {
+	SND_SOC_DAPM_MIC("Dummy MIC", NULL),
+	SND_SOC_DAPM_SPK("Dummy SPK", NULL),
+};
+
+static const struct snd_soc_dapm_route tegra_dummy_routes[] = {
+	{"Dummy SPK",		NULL,	"Dummy Playback"},
+	{"Dummy Capture",	NULL,	"Dummy MIC"},
+};
+
+static const struct snd_soc_component_driver tegra_dummy_component = {
+	.dapm_widgets		= tegra_dummy_widgets,
+	.num_dapm_widgets	= ARRAY_SIZE(tegra_dummy_widgets),
+	.dapm_routes		= tegra_dummy_routes,
+	.num_dapm_routes	= ARRAY_SIZE(tegra_dummy_routes),
+	.endianness		= 1,
+};
+
+static struct snd_soc_dai_link_component tegra_dummy_dlc = {
+	.of_node	= NULL,
+	.dai_name	= "tegra-snd-dummy-dai",
+	.name		= "sound",
+};
+
 static int tegra_audio_graph_probe(struct platform_device *pdev)
 {
 	struct tegra_audio_priv *priv;
 	struct device *dev = &pdev->dev;
 	struct snd_soc_card *card;
+	int ret;
 
 	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
 	if (!priv)
@@ -209,11 +289,31 @@ static int tegra_audio_graph_probe(struct platform_device *pdev)
 	card->probe = tegra_audio_graph_card_probe;
 
 	/* audio_graph_parse_of() depends on below */
-	card->component_chaining = 1;
+	if (!of_property_read_bool(dev->of_node, "nvidia,ahub-c2c-links")) {
+		card->component_chaining = 1;
+		priv->simple.force_dpcm = 1;
+	}
+
+	ret = devm_snd_soc_register_component(dev, &tegra_dummy_component,
+					      &tegra_dummy_dai, 1);
+	if (ret < 0) {
+		dev_err(dev, "Tegra dummy component registration fails\n");
+		return ret;
+	}
+
+	priv->simple.dummy_dlc = &tegra_dummy_dlc;
 	priv->simple.ops = &tegra_audio_graph_ops;
-	priv->simple.force_dpcm = 1;
 
-	return audio_graph_parse_of(&priv->simple, dev);
+	ret = audio_graph_parse_of(&priv->simple, dev);
+	if (ret < 0)
+		return ret;
+
+	of_platform_populate(pdev->dev.of_node, NULL, NULL, &pdev->dev);
+
+	dev_info(&pdev->dev, "Registered APE graph sound card with %s links for AHUB\n",
+		 card->component_chaining ? "DPCM" : "codec2codec");
+
+	return 0;
 }
 
 static const struct tegra_audio_cdata tegra210_data = {
-- 
2.43.0

